server/src/main/java/org/opensearch/common/metadata/MetadataParser.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.metadata;
+
+import java.io.IOException;
+
+import org.apache.lucene.codecs.CodecUtil;
+import org.apache.lucene.store.BufferedChecksumIndexInput;
+import org.apache.lucene.store.ChecksumIndexInput;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+
+/**
+ * Manages Metadata by adding version, codec in header and checksum in footer.
+ * @param <T> The type of metadata to be read/written
+ */
+public class MetadataParser<T> {
+    // This can be updated to hold a parserFactory and get relevant parsers based on the metadata versions
+    private final MetadataContentParser<T> parser;
+    private final int currentVersion;
+    private final String codec;
+
+    /**
+     * @param parser parser to read/write metadata from T
+     * @param currentVersion latest supported version of the metadata
+     * @param codec: metadata codec
+     */
+    public MetadataParser(MetadataContentParser<T> parser, int currentVersion, String codec) {
+        this.parser = parser;
+        this.currentVersion = currentVersion;
+        this.codec = codec;
+    }
```

Comment:
We need to parse the metadata based on its own version not the lastest. Each metadata can be written on a different version

Reply:
yes. This MetadataParser is responsible for both reading and writing metadata. Thinking of renaming `Parser` keyword to something else to remove this confusion.

In `readMetadata`, at ln:49, we are reading the header(`checkHeader`) which will give us the version the file was written with. Since there is just a single version as of now we have not added the logic to handle old versions. We are planning to do it separately in https://github.com/opensearch-project/OpenSearch/issues/6123 after exploring a few approaches for handling the version evolution.

After we decide on the approach, we'll add separate `MetadataContentParser` for each version and the correct one will be picked based on the version that we get while reading the metadata file header(ln:49). Refer to comment at ln:24